File: macos/Sources/Services/SettingsService.swift
import AppKit
import Foundation

/// UI-side view of the files Pob works with. The Go core (pob-core) owns
/// settings.json defaults, the src/ macros and the logs tree; this service only
/// resolves the project root, opens files in the user's editor, persists the
/// window frame and clears user files on request.
///
/// A machine has one instance and it keeps its id for good: the same
/// ~/.pob/<instance>/ directory every run, holding that instance's src/ and
/// logs/. settings.json sits above them at ~/.pob and is shared — pointing
/// ~/.pob/INSTANCE at another id starts Pob on a clean macro, on a machine
/// that is already set up.
class SettingsService {
    private let fileManager = FileManager.default

    /// ~/.pob/<instanceID>, everything this instance owns. Passed to pob-core
    /// via --instance so both sides work in the same directory.
    let instanceID: String

    /// Exclusive flock on <instanceID>/.lock, held for the process lifetime.
    /// It marks the directory as belonging to a running Pob, which is how a
    /// second Pob is detected — see claimInstance.
    ///
    /// Static, and taken exactly once. flock is per open file description,
    /// not per process, so a second open() of this same file would be refused
    /// by the lock this process already holds — Pob would find itself and
    /// conclude it was already running.
    private static var lockFD: Int32 = -1

    /// Shared project root (same for every instance in this process).
    static var projectRoot: URL {
        resolveProjectRoot(FileManager.default)
    }

    var projectRoot: URL {
        Self.resolveProjectRoot(fileManager)
    }

    private static func resolveProjectRoot(_ fileManager: FileManager) -> URL {
        // All Pob components share ~/.pob — the same root the pob CLI uses.
        let dir = fileManager.homeDirectoryForCurrentUser.appendingPathComponent(".pob")
        try? fileManager.createDirectory(at: dir, withIntermediateDirectories: true)
        return dir
    }

    var instanceDir: URL {
        projectRoot.appendingPathComponent(instanceID)
    }

    /// This instance's log — what the toolbar's .log button opens, and what
    /// pob-core writes every step and psl call to. Resolved statically and
    /// cached, because AppLogger writes to it from anywhere in the shell,
    /// including before any SettingsService has been built.
    static let instanceLogFile: URL = {
        let fileManager = FileManager.default
        let root = resolveProjectRoot(fileManager)
        let dir = root.appendingPathComponent(resolveInstanceID(fileManager, root: root))
        try? fileManager.createDirectory(at: dir, withIntermediateDirectories: true)
        return dir.appendingPathComponent("instance.log")
    }()

    var instanceLogFile: URL {
        instanceDir.appendingPathComponent("instance.log")
    }

    /// The machine's settings, shared by every instance: the API key, the model
    /// and the port are how this machine works whichever instance it is
    /// running, so moving ~/.pob/INSTANCE does not mean setting them again.
    private var settingsFile: URL {
        projectRoot.appendingPathComponent("settings.json")
    }

    /// What this instance is, rather than how it is configured: its id, the
    /// name `pob new` gave it, when it last ran, and — while it runs — the pid
    /// and control port. The Go core owns it; the window frame and its lock are
    /// the shell's entries, since where the window was and whether it can be
    /// moved are properties of the machine and not something anybody sets.
    private var instanceFile: URL {
        instanceDir.appendingPathComponent("instance.json")
    }

    /// This instance's macros. A macro of any size is written across several
    /// files — the entry point calls the pieces — so they are kept together in
    /// one directory, beside the entry point the Macro PSL button opens.
    private var srcFolder: URL {
        instanceDir.appendingPathComponent("src")
    }

    /// The entry point of this instance's Prompt Script Language program — what
    /// Record writes and Execute replays. `.macro.psl` says psl fills its
    /// slots; a `.macro` beside it is replayed without the compiler.
    private var macroFile: URL {
        srcFolder.appendingPathComponent("main.macro.psl")
    }

    private var logsFolder: URL {
        instanceDir.appendingPathComponent("logs")
    }

    init() {
        let fileManager = FileManager.default
        let root = Self.resolveProjectRoot(fileManager)

        instanceID = Self.resolveInstanceID(fileManager, root: root)
        let dir = root.appendingPathComponent(instanceID)
        try? fileManager.createDirectory(at: dir.appendingPathComponent("logs"), withIntermediateDirectories: true)
        // Normally already held — claimInstance ran at launch. This is the
        // path for anything that builds a SettingsService without it.
        Self.acquireInstanceLock(at: dir)
    }

    /// Every instance id starts with this.
    private static let instancePrefix = "pb-"

    /// The file under ~/.pob holding the machine's one instance id. Named in
    /// capitals like VERSION and SYSTEM: a marker the programs write and read,
    /// not a file to edit.
    private static let instancePointer = "INSTANCE"

    /// The machine's instance id — the same one on every run, recorded in
    /// ~/.pob/INSTANCE the first time it is worked out. This mirrors the Go
    /// core's ResolveInstanceID because either side can get there first: the
    /// shell resolves it to show in the toolbar and passes it to pob-core
    /// with --instance, but the CLI can reach ~/.pob without a shell at all.
    ///
    /// The pointer is the only thing that says which instance this machine
    /// is: with no readable one a fresh id is drawn and a new directory
    /// reserved, rather than an existing pb-* directory adopted. Deleting the
    /// file is therefore a way to start clean, and what is already there stays
    /// as history.
    private static func resolveInstanceID(_ fileManager: FileManager, root: URL) -> String {
        let pointer = root.appendingPathComponent(instancePointer)

        if let contents = try? String(contentsOf: pointer, encoding: .utf8) {
            let id = contents.trimmingCharacters(in: .whitespacesAndNewlines)
            // Anything that isn't an instance id — a truncated or hand-edited
            // file — sends us back to working it out, rather than into a
            // directory named after junk.
            if id.hasPrefix(instancePrefix), !id.contains("/") {
                return id
            }
        }

        let id = reserveInstanceID(fileManager, root: root)
        try? (id + "\n").write(to: pointer, atomically: true, encoding: .utf8)
        return id
    }

    /// Reserves a fresh `pb-<4 hex>` directory — the last two bytes of a new
    /// UID as lowercase hex, the same scheme the pico-hid firmware uses for
    /// its `ph-` board id. Shown in the toolbar beside the window buttons and
    /// used as the instance directory name, so the id on screen names the
    /// directory to look in.
    private static func reserveInstanceID(_ fileManager: FileManager, root: URL) -> String {
        while true {
            let id = instancePrefix + UUID().uuidString.suffix(4).lowercased()
            do {
                try fileManager.createDirectory(at: root.appendingPathComponent(id), withIntermediateDirectories: false)
                return id
            } catch CocoaError.fileWriteFileExists {
                continue // drawn one that already exists; draw another
            } catch {
                return id
            }
        }
    }

    /// True when this process holds the machine's instance. Anything that
    /// would drive the desktop — starting pob-core above all — has to check
    /// it, because the scene is built before the app delegate gets to refuse
    /// the launch, and a core started in that window would outlive the
    /// refusal.
    static var holdsInstanceLock: Bool { lockFD >= 0 }

    /// Claims this machine's instance for this process and reports whether
    /// it was free; false means another Pob already holds it. Called at
    /// launch, before any window is built, since only one Pob drives a
    /// desktop.
    ///
    /// Claiming and asking are the same operation on purpose. Asking first
    /// and taking it after would leave a gap for a second Pob to slip
    /// through — and, because flock belongs to the open file description
    /// rather than the process, the asking itself would collide with the lock
    /// this process had already taken.
    static func claimInstance() -> Bool {
        let fileManager = FileManager.default
        let root = resolveProjectRoot(fileManager)
        let dir = root.appendingPathComponent(resolveInstanceID(fileManager, root: root))
        try? fileManager.createDirectory(at: dir, withIntermediateDirectories: true)
        return acquireInstanceLock(at: dir)
    }

    deinit {
        // The lock is the process's, not this object's: SwiftUI can release
        // and rebuild a @StateObject while the app runs, and dropping the
        // lock there would let a second Pob in. The OS releases it on exit.
    }

    private func serializeJSON(_ object: Any) -> String? {
        guard let data = try? JSONSerialization.data(withJSONObject: object, options: [.prettyPrinted, .sortedKeys, .withoutEscapingSlashes]),
              var string = String(data: data, encoding: .utf8) else { return nil }
        string = string.replacingOccurrences(of: "\" : ", with: "\": ")
        return string
    }

    func getEditor() -> String {
        loadJSON(key: "editor") as? String ?? "system"
    }

    func getTerminal() -> String {
        loadJSON(key: "terminal") as? String ?? "system"
    }

    /// The frame the window was last left at, from instance.json — or from
    /// settings.json, where it used to be kept. Both are read because either
    /// side can get here first on the run that moves it: pob-core carries the
    /// frame over at startup, and this is called as the window is built. The
    /// fallback goes quiet once core has run, which drops it from settings.json.
    func getWindowFrame() -> NSRect? {
        readFrame(loadJSON(instanceFile)) ?? readFrame(loadJSON(settingsFile))
    }

    private func readFrame(_ json: [String: Any]) -> NSRect? {
        guard let x = json["window_x"] as? Double,
              let y = json["window_y"] as? Double,
              let w = json["window_width"] as? Double,
              let h = json["window_height"] as? Double else { return nil }
        return NSRect(x: x, y: y, width: w, height: h)
    }

    func saveWindowFrame(_ frame: NSRect) {
        var json = loadJSON(instanceFile)
        json["window_x"] = Double(frame.origin.x)
        json["window_y"] = Double(frame.origin.y)
        json["window_width"] = Double(frame.size.width)
        json["window_height"] = Double(frame.size.height)
        writeInstanceFile(json)
    }

    /// Whether the window was left locked, from instance.json. It belongs
    /// beside the frame: the lock is what holds the frame to its size and to
    /// what it frames, so a run that restored the frame but not the lock would
    /// come back loose — and a window locked to hold a macro's coordinates
    /// would have to be locked again by hand every launch. Unlocked is the
    /// answer for an instance that has never recorded one.
    func getWindowLocked() -> Bool {
        loadJSON(instanceFile)["is_locked"] as? Bool ?? false
    }

    func saveWindowLocked(_ locked: Bool) {
        var json = loadJSON(instanceFile)
        json["is_locked"] = locked
        writeInstanceFile(json)
    }

    /// Whether the window was left passing clicks through, from instance.json.
    /// It belongs beside the lock for the same reason: an instance set up to
    /// sit over the app it drives comes back sitting over it, instead of
    /// swallowing the first clicks meant for what is underneath until the
    /// button is pressed again. On is the answer for an instance that has never
    /// recorded one — the overlay's resting state.
    func getClickThrough() -> Bool {
        loadJSON(instanceFile)["is_click_through"] as? Bool ?? true
    }

    func saveClickThrough(_ enabled: Bool) {
        var json = loadJSON(instanceFile)
        json["is_click_through"] = enabled
        writeInstanceFile(json)
    }

    /// Writes instance.json whole, so everything already in it — the id, the
    /// name, the times the core keeps — survives the shell's own entries.
    private func writeInstanceFile(_ json: [String: Any]) {
        if let string = serializeJSON(json) {
            try? string.write(to: instanceFile, atomically: true, encoding: .utf8)
        }
    }

    func openSettingsFile() {
        openWithEditor(settingsFile)
    }

    /// Opens the entry point rather than the src/ directory around it: what
    /// someone reaches for the Macro PSL button to do is edit the macro, and
    /// that starts at main.macro.psl — the rest of src/ is a `call()` away in
    /// the editor that is now already open on it.
    func openMacroFile() {
        ensureSrcFolder()
        if !fileManager.fileExists(atPath: macroFile.path) {
            fileManager.createFile(atPath: macroFile.path, contents: Data())
        }
        openWithEditor(macroFile)
    }

    func getMacro() -> String {
        (try? String(contentsOf: macroFile, encoding: .utf8)) ?? ""
    }

    func clearMacro() {
        ensureSrcFolder()
        try? "".write(to: macroFile, atomically: true, encoding: .utf8)
    }

    /// The core makes src/ at startup; this is for the writes that can land
    /// before it has, and costs a stat on a directory that is already there.
    private func ensureSrcFolder() {
        try? fileManager.createDirectory(at: srcFolder, withIntermediateDirectories: true)
    }

    /// Appends one action line to main.macro.psl (same format as the Go core's
    /// AppendToMacro). Only called while no session is executing, so it never
    /// races with the Go core's own appends.
    func appendToMacro(_ line: String) {
        ensureSrcFolder()
        var content = getMacro()
        if !content.isEmpty, !content.hasSuffix("\n") { content += "\n" }
        content += line + "\n"
        try? content.write(to: macroFile, atomically: true, encoding: .utf8)
    }

    /// Removes the last non-empty line if it equals `expected`; used by the
    /// recorder to upgrade a click() into a doubleClick(). Returns whether a
    /// line was removed.
    func removeLastMacroLine(ifMatches expected: String) -> Bool {
        var lines = getMacro().components(separatedBy: "\n")
        while let last = lines.last, last.trimmingCharacters(in: .whitespaces).isEmpty {
            lines.removeLast()
        }
        guard lines.last == expected else { return false }
        lines.removeLast()
        var content = lines.joined(separator: "\n")
        if !content.isEmpty { content += "\n" }
        try? content.write(to: macroFile, atomically: true, encoding: .utf8)
        return true
    }

    /// Takes the flock, or reports that someone else has it. Already holding
    /// it counts as success — this process is the someone else.
    @discardableResult
    private static func acquireInstanceLock(at instanceDir: URL) -> Bool {
        if lockFD >= 0 { return true }
        let fd = open(instanceDir.appendingPathComponent(".lock").path, O_CREAT | O_RDWR, 0o644)
        // A lock file that won't open is a broken ~/.pob, not a second Pob.
        // Start anyway rather than refuse over something unrelated.
        guard fd >= 0 else { return true }
        // Non-blocking: a held lock is an answer, not something to wait for.
        if flock(fd, LOCK_EX | LOCK_NB) != 0 {
            close(fd)
            return false
        }
        lockFD = fd
        return true
    }

    func openLogsFolder() {
        try? fileManager.createDirectory(at: logsFolder, withIntermediateDirectories: true)
        NSWorkspace.shared.open(logsFolder)
    }

    /// The instance log, not the app log: what someone reaches for a log for is
    /// what a run did, and that is written here in full. app.log keeps only the
    /// app and its instances starting, stopping and failing.
    func openInstanceLog() {
        let log = instanceLogFile
        if !fileManager.fileExists(atPath: log.path) {
            fileManager.createFile(atPath: log.path, contents: Data())
        }
        openWithEditor(log)
    }

    /// The app log, behind ⌥-click on the same button: when the question is
    /// about the app rather than a run — which instances started, which
    /// stopped, what failed — the answer is across instances, not inside one.
    func openAppLog() {
        let log = projectRoot.appendingPathComponent("app.log")
        if !fileManager.fileExists(atPath: log.path) {
            fileManager.createFile(atPath: log.path, contents: Data())
        }
        openWithEditor(log)
    }

    private func openWithEditor(_ url: URL) {
        let editor = getEditor()
        let process = Process()
        switch editor {
        case "vscode":
            process.launchPath = "/usr/bin/open"
            process.arguments = ["-a", "Visual Studio Code", url.path]
        case "zed":
            process.launchPath = "/usr/bin/open"
            process.arguments = ["-a", "Zed", url.path]
        case "sublime_text":
            process.launchPath = "/usr/bin/open"
            process.arguments = ["-a", "Sublime Text", url.path]
        case "vim":
            process.launchPath = "/usr/bin/osascript"
            let escaped = url.path.replacingOccurrences(of: "\\", with: "\\\\").replacingOccurrences(of: "\"", with: "\\\"")
            let cmd = "vim \\\"\(escaped)\\\""
            if getTerminal() == "iterm2" {
                process.arguments = ["-e", "tell application \"iTerm\" to create window with default profile command \"vim \\\"\(escaped)\\\"\"", "-e", "tell application \"iTerm\" to activate"]
            } else {
                process.arguments = ["-e", "tell application \"Terminal\" to do script \"\(cmd)\"", "-e", "tell application \"Terminal\" to activate"]
            }
        default: // "system"
            openWithSystemEditor(url)
            return
        }

        // An editor settings.json names may not be installed here: the setting
        // travels between machines, the application does not. `open -a` and
        // osascript both report that in their exit status, which is the only
        // sign the file never appeared, so hand it to the system editor rather
        // than leave the toolbar button looking dead.
        process.terminationHandler = { [weak self] finished in
            guard finished.terminationStatus != 0 else { return }
            AppLogger.log("Settings: \(editor) could not open \(url.path) — using the system editor")
            DispatchQueue.main.async { self?.openWithSystemEditor(url) }
        }
        do {
            try process.run()
        } catch {
            AppLogger.error("Settings: cannot run \(process.launchPath ?? editor): \(error.localizedDescription)")
            openWithSystemEditor(url)
        }
    }

    /// The file in whatever is registered for it, TextEdit behind that. The
    /// last resort on macOS, and one that always has something behind it —
    /// unlike Linux, where xdg-open can come up empty.
    private func openWithSystemEditor(_ url: URL) {
        let process = Process()
        process.launchPath = "/usr/bin/open"
        process.arguments = ["-t", url.path]
        try? process.run()
    }

    private func loadJSON(key: String) -> Any? {
        loadJSON(settingsFile)[key]
    }

    private func loadJSON(_ file: URL) -> [String: Any] {
        guard let data = try? Data(contentsOf: file),
              let json = try? JSONSerialization.jsonObject(with: data) as? [String: Any] else { return [:] }
        return json
    }
}

File: win/src/Services/SettingsService.cs
// UI-side view of the files Pob works with, mirroring the macOS/Linux
// SettingsService. The Go core owns settings.json defaults, the src/ macros
// and the logs tree; this service only resolves the project root,
// opens files in the user's editor, persists the window frame and clears
// user files on request.
//
// What an instance owns lives under ~/.pob/<instance>/; settings.json sits
// above them at ~/.pob and is shared — pointing ~/.pob/INSTANCE at another id
// starts Pob on a clean macro, on a machine already set up.
using System.Diagnostics;
using System.IO;
using System.Text.Json;
using System.Text.Json.Nodes;

namespace Pob.Services;

public static class SettingsService
{
    // ── project root ────────────────────────────────────────────────────────

    private static string? _root;

    public static string ProjectRoot => _root ??= ComputeRoot();

    private static string ComputeRoot()
    {
        // All Pob components share ~/.pob — the same root the pob CLI uses.
        string root = Path.Combine(
            Environment.GetFolderPath(Environment.SpecialFolder.UserProfile), ".pob");
        Directory.CreateDirectory(root);
        return root;
    }

    private static string RootPath(string name) => Path.Combine(ProjectRoot, name);

    // ── instance directory ──────────────────────────────────────────────────

    private static string? _instanceId;

    /// <summary>
    /// ~/.pob/&lt;InstanceId&gt;, everything this instance owns: its
    /// src/ and logs/. Passed to pob-core via
    /// --instance.
    /// </summary>
    public static string InstanceId => _instanceId ??= AllocateInstance();

    /// <summary>
    /// Exclusive handle on &lt;InstanceId&gt;/.lock, held for the process
    /// lifetime. It marks the directory as belonging to a running Pob, which
    /// is how a second Pob is detected — see ClaimInstance.
    ///
    /// Opened exactly once. The share mode is enforced per handle, not per
    /// process, so opening this same file a second time would be refused by
    /// the handle this process already holds — Pob would find itself and
    /// conclude it was already running.
    /// </summary>
    private static FileStream? _instanceLock;

    private const string InstancePrefix = "pb-";

    /// <summary>
    /// The file under ~/.pob holding the machine's one instance id. Named in
    /// capitals like VERSION and SYSTEM: a marker the programs write and read,
    /// not a file to edit.
    /// </summary>
    private const string InstancePointer = "INSTANCE";

    private static string AllocateInstance()
    {
        string id = ResolveInstanceId(ProjectRoot);
        string dir = Path.Combine(ProjectRoot, id);
        Directory.CreateDirectory(Path.Combine(dir, "logs"));
        AcquireInstanceLock(dir);
        return id;
    }

    /// <summary>~/.pob/&lt;instance&gt;, everything this instance owns.</summary>
    public static string InstanceDir => Path.Combine(ProjectRoot, InstanceId);

    private static string InstancePath(string name) => Path.Combine(InstanceDir, name);

    /// <summary>
    /// This instance's macros, ~/.pob/&lt;instance&gt;/src. A macro of any size
    /// is written across several files — the entry point calls the pieces — so
    /// they are kept together in one directory, which is what the Macro PSL
    /// button opens.
    /// </summary>
    private static string SrcDir => InstancePath("src");

    /// <summary>
    /// The entry point of this instance's macro. `.macro.psl` says psl fills
    /// its slots; a `.macro` beside it is replayed without the compiler.
    /// </summary>
    private static string MacroPath => Path.Combine(SrcDir, "main.macro.psl");

    // The core makes src/ at startup; this is for the writes that can land
    // before it has, and costs a stat on a directory that is already there.
    private static void EnsureSrcDir() => Directory.CreateDirectory(SrcDir);

    /// <summary>
    /// The machine's instance id — the same one on every run, recorded in
    /// ~/.pob/INSTANCE the first time it is worked out. This mirrors the Go
    /// core's ResolveInstanceID because either side can get there first: the
    /// shell resolves it to show in the toolbar and passes it to pob-core
    /// with --instance, but the CLI can reach ~/.pob without a shell at all.
    ///
    /// The pointer is the only thing that says which instance this machine
    /// is: with no readable one a fresh id is drawn and a new directory
    /// reserved, rather than an existing pb-* directory adopted. Deleting the
    /// file is therefore a way to start clean, and what is already there stays
    /// as history.
    /// </summary>
    private static string ResolveInstanceId(string root)
    {
        string pointer = RootPath(InstancePointer);
        try
        {
            string id = File.ReadAllText(pointer).Trim();
            // Anything that isn't an instance id — a truncated or hand-edited
            // file — sends us back to working it out, rather than into a
            // directory named after junk.
            if (id.StartsWith(InstancePrefix, StringComparison.Ordinal) &&
                id.IndexOfAny([Path.DirectorySeparatorChar, Path.AltDirectorySeparatorChar]) < 0)
            {
                return id;
            }
        }
        catch (IOException)
        {
        }
        catch (UnauthorizedAccessException)
        {
        }

        string resolved = ReserveInstanceId(root);
        try
        {
            File.WriteAllText(pointer, resolved + Environment.NewLine);
        }
        catch (IOException)
        {
        }
        return resolved;
    }

    /// <summary>
    /// Reserves a fresh pb-&lt;4 hex&gt; directory — the last two bytes of a new
    /// UID as lowercase hex, the same scheme the pico-hid firmware uses for
    /// its ph- board id. The toolbar shows it beside the window buttons, so
    /// the id on screen names the directory to look in.
    /// </summary>
    private static string ReserveInstanceId(string root)
    {
        string id = NewInstanceId();
        while (Directory.Exists(Path.Combine(root, id))) id = NewInstanceId();
        return id;
    }

    private static string NewInstanceId() => InstancePrefix + Guid.NewGuid().ToString("N")[28..];

    /// <summary>
    /// Claims this machine's instance for this process and reports whether it
    /// was free; false means another Pob already holds it. Called at launch,
    /// before any window is built, since only one Pob drives a desktop.
    ///
    /// Claiming and asking are the same operation on purpose. Asking first
    /// and taking it after would leave a gap for a second Pob to slip
    /// through — and, because the share mode belongs to the handle rather
    /// than the process, the asking itself would collide with the handle this
    /// process had already opened.
    /// </summary>
    public static bool ClaimInstance()
    {
        string dir = Path.Combine(ProjectRoot, ResolveInstanceId(ProjectRoot));
        Directory.CreateDirectory(dir);
        return AcquireInstanceLock(dir);
    }

    /// <summary>
    /// The machine's settings, shared by every instance: the API key, the model
    /// and the port are how this machine works whichever instance it is
    /// running, so moving ~/.pob/INSTANCE does not mean setting them again.
    /// </summary>
    private static string SettingsFilePath() => Path.Combine(ProjectRoot, "settings.json");

    /// <summary>
    /// What this instance is rather than how it is configured: its id, the name
    /// `pob new` gave it, when it last ran and, while it runs, the pid and
    /// control port. The Go core owns the file; the window frame is the shell's
    /// one entry, since where the window was is a property of the machine
    /// rather than something anybody sets.
    /// </summary>
    private static string InstanceFilePath() => InstancePath("instance.json");

    // ── settings.json helpers ───────────────────────────────────────────────

    private static JsonObject? LoadSettings() => LoadJson(SettingsFilePath());

    private static JsonObject? LoadInstance() => LoadJson(InstanceFilePath());

    private static JsonObject? LoadJson(string path)
    {
        try
        {
            return JsonNode.Parse(File.ReadAllText(path)) as JsonObject;
        }
        catch
        {
            return null;
        }
    }

    private static string LoadStringKey(string key, string fallback)
    {
        JsonObject? obj = LoadSettings();
        if (obj != null && obj.TryGetPropertyValue(key, out JsonNode? node) && node is JsonValue v &&
            v.TryGetValue(out string? s) && s != null)
            return s;
        return fallback;
    }

    /// <summary>
    /// The frame comes from instance.json, or from settings.json where it used
    /// to be kept. Both are read because either side can get here first on the
    /// run that moves it: pob-core carries the frame over at startup, and this
    /// is called as the window is built. The fallback goes quiet once core has
    /// run, which drops the keys from settings.json.
    /// </summary>
    public static bool GetWindowFrame(out int x, out int y, out int w, out int h)
    {
        if (ReadFrame(LoadInstance(), out x, out y, out w, out h)) return true;
        return ReadFrame(LoadSettings(), out x, out y, out w, out h);
    }

    private static bool ReadFrame(JsonObject? obj, out int x, out int y, out int w, out int h)
    {
        x = y = 0;
        w = 600;
        h = 400;
        if (obj == null) return false;
        if (!TryGetInt(obj, "window_x", out x) || !TryGetInt(obj, "window_y", out y) ||
            !TryGetInt(obj, "window_width", out w) || !TryGetInt(obj, "window_height", out h))
            return false;
        return true;
    }

    private static bool TryGetInt(JsonObject obj, string key, out int value)
    {
        value = 0;
        if (!obj.TryGetPropertyValue(key, out JsonNode? node) || node is not JsonValue v) return false;
        if (v.TryGetValue(out double d))
        {
            value = (int)d;
            return true;
        }
        return false;
    }

    public static void SaveWindowFrame(int x, int y, int w, int h)
    {
        // Preserve every existing key, only replace the frame values.
        JsonObject obj = LoadInstance() ?? new JsonObject();
        obj["window_x"] = (double)x;
        obj["window_y"] = (double)y;
        obj["window_width"] = (double)w;
        obj["window_height"] = (double)h;
        WriteInstance(obj);
    }

    /// <summary>
    /// Whether the window was left locked. It belongs beside the frame: the
    /// lock is what keeps the frame still, so a run that restored the frame but
    /// not the lock would come back movable — and a window locked to hold a
    /// macro's coordinates would have to be locked again by hand every launch.
    /// False for an instance that has never recorded one.
    /// </summary>
    public static bool GetWindowLocked()
    {
        JsonObject? obj = LoadInstance();
        if (obj == null) return false;
        if (!obj.TryGetPropertyValue("is_locked", out JsonNode? node) || node is not JsonValue v) return false;
        return v.TryGetValue(out bool locked) && locked;
    }

    public static void SaveWindowLocked(bool locked)
    {
        JsonObject obj = LoadInstance() ?? new JsonObject();
        obj["is_locked"] = locked;
        WriteInstance(obj);
    }

    /// <summary>
    /// Whether the window was left passing clicks through. It belongs beside
    /// the lock for the same reason: an instance set up to sit over the app it
    /// drives comes back sitting over it, instead of swallowing the first
    /// clicks meant for what is underneath until the button is pressed again.
    /// True for an instance that has never recorded one — the overlay's resting
    /// state.
    /// </summary>
    public static bool GetClickThrough()
    {
        JsonObject? obj = LoadInstance();
        if (obj == null) return true;
        if (!obj.TryGetPropertyValue("is_click_through", out JsonNode? node) || node is not JsonValue v)
            return true;
        if (!v.TryGetValue(out bool on)) return true;
        return on;
    }

    public static void SaveClickThrough(bool on)
    {
        JsonObject obj = LoadInstance() ?? new JsonObject();
        obj["is_click_through"] = on;
        WriteInstance(obj);
    }

    // Writes instance.json whole, so everything already in it — the id, the
    // name, the times and the port the core keeps — survives the shell's own
    // entries.
    private static void WriteInstance(JsonObject obj)
    {
        try
        {
            File.WriteAllText(InstanceFilePath(),
                obj.ToJsonString(new JsonSerializerOptions { WriteIndented = true }));
        }
        catch (IOException)
        {
        }
    }

    // ── opening files ───────────────────────────────────────────────────────

    private static void SpawnDetached(string fileName, params string[] args)
    {
        try
        {
            var psi = new ProcessStartInfo(fileName)
            {
                UseShellExecute = false,
                CreateNoWindow = true,
            };
            foreach (string a in args) psi.ArgumentList.Add(a);
            Process.Start(psi);
        }
        catch (Exception e)
        {
            AppLogger.Error($"Failed to launch {fileName}: {e.Message}");
        }
    }

    // Whether a command is on the PATH, the same lookup cmd.exe does, PATHEXT
    // included. Worth asking first: `cmd /c code` exits quietly when VS Code
    // isn't installed, and a quiet exit is exactly what leaves the toolbar
    // button looking dead.
    private static bool ExistsOnPath(string command)
    {
        string exts = Environment.GetEnvironmentVariable("PATHEXT") ?? ".COM;.EXE;.BAT;.CMD";
        string dirs = Environment.GetEnvironmentVariable("PATH") ?? "";
        foreach (string dir in dirs.Split(Path.PathSeparator))
        {
            if (string.IsNullOrWhiteSpace(dir)) continue;
            foreach (string ext in exts.Split(';'))
            {
                // A PATH entry with invalid characters is one bad entry, not a
                // reason to stop looking through the rest.
                try
                {
                    if (File.Exists(Path.Combine(dir.Trim(), command + ext))) return true;
                }
                catch (ArgumentException)
                {
                }
            }
        }
        return false;
    }

    private static void OpenWithEditor(string path)
    {
        string editor = LoadStringKey("editor", "system");

        switch (editor)
        {
            case "vscode":
                // VS Code's CLI is code.cmd — go through cmd so PATH lookup works.
                if (ExistsOnPath("code")) { SpawnDetached("cmd.exe", "/c", "code", path); return; }
                break;
            case "zed":
                if (ExistsOnPath("zed")) { SpawnDetached("cmd.exe", "/c", "zed", path); return; }
                break;
            case "sublime_text":
                if (ExistsOnPath("subl")) { SpawnDetached("cmd.exe", "/c", "subl", path); return; }
                break;
            case "vim":
                if (ExistsOnPath("vim"))
                {
                    string terminal = LoadStringKey("terminal", "system");
                    if ((terminal == "wt" || terminal == "windows_terminal") && ExistsOnPath("wt"))
                        SpawnDetached("wt.exe", "vim", path);
                    else // "system": a plain console window
                        SpawnDetached("cmd.exe", "/c", "start", "vim", path);
                    return;
                }
                break;
        }

        // "system", or an editor settings.json names that isn't installed here
        // — the setting travels between machines, the editor does not.
        if (editor != "system")
            AppLogger.Log($"Settings: {editor} is not installed — opening {path} with the system editor");
        OpenWithSystemEditor(path);
    }

    // The file-type association (the xdg-open equivalent), with Notepad behind
    // it for .log & friends that have none.
    private static void OpenWithSystemEditor(string path)
    {
        try
        {
            Process.Start(new ProcessStartInfo(path) { UseShellExecute = true });
        }
        catch (Exception e)
        {
            AppLogger.Log($"Settings: nothing is registered for {path} ({e.Message}) — using Notepad");
            SpawnDetached("notepad.exe", path);
        }
    }

    private static void EnsureFile(string path)
    {
        if (!File.Exists(path)) File.WriteAllText(path, "");
    }

    public static void OpenSettingsFile()
    {
        string path = SettingsFilePath();
        EnsureFile(path);
        OpenWithEditor(path);
    }

    // Opens the whole src/ directory rather than the entry point alone: what
    // someone reaches for the Macro PSL button to do is edit the macro, and a
    // macro is the set of files, not the one that happens to be called first.
    public static void OpenSrcFolder()
    {
        EnsureSrcDir();
        SpawnDetached("explorer.exe", SrcDir);
    }

    // The instance log, not the app log: what someone reaches for a log for is
    // what a run did, and that is written here in full. app.log keeps only the
    // app and its instances starting, stopping and failing.
    public static void OpenInstanceLog()
    {
        string path = InstancePath("instance.log");
        EnsureFile(path);
        OpenWithEditor(path);
    }

    public static void OpenLogsFolder()
    {
        string path = InstancePath("logs");
        Directory.CreateDirectory(path);
        SpawnDetached("explorer.exe", path);
    }

    // ── file contents / clearing ────────────────────────────────────────────

    public static string GetMacro()
    {
        try
        {
            return File.ReadAllText(MacroPath);
        }
        catch
        {
            return "";
        }
    }

    // Appends one action line, keeping the file newline-terminated. Read-
    // modify-write like the macOS shell's appendToMacro: the macro is small
    // and the core is the only other writer.
    public static void AppendToMacro(string line)
    {
        string content = GetMacro();
        if (content.Length > 0 && !content.EndsWith("\n")) content += "\n";
        try
        {
            EnsureSrcDir();
            File.WriteAllText(MacroPath, content + line + "\n");
        }
        catch (IOException)
        {
        }
    }

    public static void ClearMacro() => TryTruncate(MacroPath);

    /// <summary>
    /// Takes the lock handle, or reports that someone else has it. Already
    /// holding it counts as success — this process is the someone else.
    /// </summary>
    private static bool AcquireInstanceLock(string instanceDir)
    {
        if (_instanceLock != null) return true;
        try
        {
            _instanceLock = new FileStream(Path.Combine(instanceDir, ".lock"),
                FileMode.OpenOrCreate, FileAccess.ReadWrite, FileShare.None);
            return true;
        }
        catch (IOException)
        {
            // A sharing violation is another Pob. Anything else here is a
            // broken ~/.pob, but both look the same from out here and
            // refusing to start is the safer reading of the two.
            return false;
        }
        catch (UnauthorizedAccessException)
        {
            // Permissions, not a second Pob. Start anyway rather than refuse
            // over something unrelated.
            return true;
        }
    }

    private static void TryTruncate(string path)
    {
        try
        {
            File.WriteAllText(path, "");
        }
        catch (IOException)
        {
        }
    }
}
